README
Coyote Task Manager

Copyright (C) 2000-2019 Isaque Galdino de Araujo.
Written by Isaque Galdino de Araujo <igaldino@gmail.com>

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
02111-1307, USA.

HOW TO BUILD IT

Install libdb_cxx and libfltk development files

Download the code

$ tar xzf coyotetm-<latest-version>.tgz
$ cd coyotetm
$ ./configure
$ make
$ sudo make install

HOW TO EXECUTE IT

$ /usr/local/bin/coyotetm

It will create ~/coyotetm directory for database files

You have three arguments:
--dump
   it will dump the data in a text format to you stdout
--reindex
   if for whatever reason the application crash, it might
   damage your index, so just reindexit
--version
   application version



ChangeLog
2021-04-19 v1.1

Fix flathub compliance issues
Fix appdata screenshots URLs

2021-04-18 v1.0

I believe it's time!

2020-08-18 v0.9

Moved to Github;
Changed to meson;
Added flatpak support; and
Added appdata.

2015-10-15 v0.8

Fixed compilation warnings when including xpm files;
Removed borders from all menubar and toolbar buttons;
Added code to persist chosen color scheme and selected show all tasks option across sessions;
Added 4 color schemes, 2 light (the default based on Adwaita palette and another one based on Ambiance palette) and 2 dark (the previous dark one and the new one based on Numix palette);
Added read-only mode for notes fields, so use can select text. Before it was not possible;
Cancel pending work code refactoring done.

2015-08-28 v0.7

Fixed warnings about using variables in fl_choice;
Added support for ambience color palette;
Changed icon handling to support fltk-1.3.* (before it was supporting only fltk-1.3.3 and above which is not the version in Ubuntu/Linux Mint);
Changed internal image array to Fl_Image instead of Fl_Pixmap, so we can handle Fl_RGB_Image (for icon);
Implemented support for multiple color schemes;
Removal of enumeration names;
Fixed mouse click focus on list items;
Enabled enter key to open list items;
Added sort to person and project uis;
Fix list view so variable column can be at the beginning of the the list;
Prevent saving history which has some pending error to fix;
Won't allow when date to be earlier than begin task date when editing history;
Added remove button to edit history, project and person windows;
Prevent saving tasks which has some pending error to fix;
Won't allow end date to be earlier than begin date when editing task;
Added mark canceled and completed buttons to task edit window;
Moved task description to the beginning of the list again;
Replaced "Cancel" to "Mark Canceled" task in task view;
Improved messages for remove person, project and history and to mark task canceled and completed information windows;
Improved read-only input fields look to match with inactive;
Added option --dev;
Changed task count in task view;
Narrowed list view borders;
Removed checking for message in notify;
Added key up and down in list views;
Create a list base class;
Fix directory created window;
Add shortcuts to ok and cancel button;
Fix spinning cursor in gnome; and
Change File / Edit to File / Open.

2015-07-13 v0.6

Added a task combobox in history view, so you'll know which task that history is refering too;
Changed remove button in history view to only remove history of tasks in progress. It will ask to open the task view otherwise;
Small fix to some parts where CoyoteTM was displaying Coyote TM;
Fixed spacing between menus and shortcuts; and
Merged canceled and completed view menu checkboxes to only "show all" now.

2015-07-12 v0.5

Fixed task list view buggy multiselection by dropping FL_DRAG event
Changed task list view sort to sort by category and then sort by status
Changed don't start task behaviour to open history but in read-only mode
Added read-only mode for task edit window when status is: ON HOLD, COMPLETED or CANCELED
Added history view
Added desktop file for DE integration, besides application icons
Improved about window
Added --version command line parameter

2015-06-14 v0.4

Move count to toolbar
Add shortcuts
Start a task when a new follow up is created
Check for project and person before saving
Fix you still have "working" in progress
Add dark color schema
Apply square boxes to all widgets
Fix notes boxes to FL_DOWN_BOX
Fix take_focus
Fix task list sort
Fix buttons space in edit windows
Add minimum size for windows
Update titles when changes description/name
Change person and project list column label 1 to name
Fix person list resize
Fix follow up sort
Fix task list in edit windows
Reduce font size
Don't deactive project neither person
Add validation to ProjectUI
Add validation to PersonEditUI
Auto select project and person if there is only one

2014-07-24 v0.3

fixed few bugs in the fltk version

2014-04-29 v0.2

will be last fltk version



TaskUI.cxx
// TaskUI.h <<GUI>>
//
// This file contains the TaskUI class implementation.
//
// Copyright (C) 2000-2015 Isaque Galdino de Araujo.
// Written by Isaque Galdino de Araujo <igaldino@gmail.com>
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
// 02111-1307, USA.

#include <TaskUI.h>

// default window width, height, field height, field gap and button width
#define DWH 450

// callbacks
enum
{
   NO_CB,
   ABOUT_CB,
   ADD_CB,
   COLOR_AMBIANCE_CB,
   COLOR_DARK_CB,
   COLOR_DEFAULT_CB,
   COLOR_NUMIX_CB,
   EDIT_CB,
   LIST_CB,
   MARK_CANCELED_CB,
   MARK_COMPLETED_CB,
   QUIT_CB,
   SHOW_ALL_CB,
   VIEW_HISTORY_CB,
   VIEW_PERSON_CB,
   VIEW_PROJECT_CB
};

// menu items
static Fl_Menu_Item menu[]=
{
   {"&File", 0, 0, 0, FL_SUBMENU, 0, 0, 0, 0},
      {"&New             ", FL_COMMAND+'n', (Fl_Callback*)TaskUI::all_cb, (void*)ADD_CB,  0,               0, 0, 0, 0},
      {"&Open            ", FL_COMMAND+'o', (Fl_Callback*)TaskUI::all_cb, (void*)EDIT_CB, FL_MENU_DIVIDER, 0, 0, 0, 0},
      {"&Quit            ", FL_COMMAND+'q', (Fl_Callback*)TaskUI::all_cb, (void*)QUIT_CB, 0,               0, 0, 0, 0},
      {0, 0, 0, 0, 0, 0, 0, 0, 0},
   {"&Edit", 0, 0, 0, FL_SUBMENU, 0, 0, 0, 0},
      {"Mark Completed   ", FL_COMMAND+'k', (Fl_Callback*)TaskUI::all_cb, (void*)MARK_COMPLETED_CB, 0, 0, 0, 0, 0},
      {"Mark Canceled    ", FL_COMMAND+'x', (Fl_Callback*)TaskUI::all_cb, (void*)MARK_CANCELED_CB,  0, 0, 0, 0, 0},
      {0, 0, 0, 0, 0, 0, 0, 0, 0},
   {"&View", 0, 0, 0, FL_SUBMENU, 0, 0, 0, 0},
      {"&Project         ", FL_COMMAND+'2', (Fl_Callback*)TaskUI::all_cb, (void*)VIEW_PROJECT_CB, 0,                                 0, 0, 0, 0},
      {"Per&son          ", FL_COMMAND+'3', (Fl_Callback*)TaskUI::all_cb, (void*)VIEW_PERSON_CB,  0,                                 0, 0, 0, 0},
      {"&History         ", FL_COMMAND+'4', (Fl_Callback*)TaskUI::all_cb, (void*)VIEW_HISTORY_CB, FL_MENU_DIVIDER,                   0, 0, 0, 0},
      {"Show &All        ", FL_COMMAND+'a', (Fl_Callback*)TaskUI::all_cb, (void*)SHOW_ALL_CB,     FL_MENU_TOGGLE | FL_MENU_DIVIDER,  0, 0, 0, 0},
      {"&Colors", 0, 0, 0, FL_SUBMENU, 0, 0, 0, 0},
         {"&Default         ", 0, (Fl_Callback*)TaskUI::all_cb, (void*)COLOR_DEFAULT_CB,  FL_MENU_RADIO | FL_MENU_VALUE, 0, 0, 0, 0},
         {"Dar&k            ", 0, (Fl_Callback*)TaskUI::all_cb, (void*)COLOR_DARK_CB,     FL_MENU_RADIO,                 0, 0, 0, 0},
         {"&Ambiance        ", 0, (Fl_Callback*)TaskUI::all_cb, (void*)COLOR_AMBIANCE_CB, FL_MENU_RADIO,                 0, 0, 0, 0},
         {"&Numix           ", 0, (Fl_Callback*)TaskUI::all_cb, (void*)COLOR_NUMIX_CB,    FL_MENU_RADIO,                 0, 0, 0, 0},
         {0, 0, 0, 0, 0, 0, 0, 0, 0},
      {0, 0, 0, 0, 0, 0, 0, 0, 0},
   {"&Help", 0, 0, 0, FL_SUBMENU, 0, 0, 0, 0},
      {"&About CoyoteTM  ", 0, (Fl_Callback*)TaskUI::all_cb, (void*)ABOUT_CB, 0, 0, 0, 0, 0},
      {0, 0, 0, 0, 0, 0, 0, 0, 0},
   {0, 0, 0, 0, 0, 0, 0, 0, 0}
};

TaskUI::TaskUI(int argc, char* argv[]):
      dao_(TaskDAO::get_dao()), ui_(UI::get_ui()),
      images_(ImageDB::get_image_db()), window_(0), menubar_(0), list_(0), sb_(0)
{
   dao_.attach(*this);

   // Window
   window_=new Fl_Window(UI::default_window_width, DWH, "CoyoteTM");
   window_->begin();

   // Menubar
   menubar_=new Fl_Menu_Bar(0, 0, UI::default_window_width, UI::default_field_height);
   menubar_->menu(menu);
   menubar_->box(FL_NO_BOX);

   // Toolbar
   Fl_Group* toolbar=new Fl_Group(0, menubar_->y()+menubar_->h()+UI::default_field_gap, UI::default_window_width, UI::default_field_height);
   toolbar->begin();

   // Toolbar / New Task
   Fl_Button* button=0;
   button=new Fl_Button(0, toolbar->y(), UI::default_field_height, UI::default_field_height);
   button->tooltip("New Task");
   button->image(images_.get(ImageDB::ADD));
   button->align(FL_ALIGN_IMAGE_NEXT_TO_TEXT);
   button->callback(all_cb, (void*)ADD_CB);
   button->box(FL_NO_BOX);
   button->down_box(FL_UP_BOX);

   // Toolbar / Edit Task
   button=new Fl_Button(button->x()+button->w()+UI::default_field_gap, toolbar->y(), UI::default_field_height, UI::default_field_height);
   button->tooltip("Edit Task");
   button->image(images_.get(ImageDB::EDIT));
   button->align(FL_ALIGN_IMAGE_NEXT_TO_TEXT);
   button->callback(all_cb, (void*)EDIT_CB);
   button->box(FL_NO_BOX);
   button->down_box(FL_UP_BOX);

   // Toolbar / Mark Completed
   button=new Fl_Button(button->x()+button->w()+UI::default_field_gap, toolbar->y(), UI::default_field_height, UI::default_field_height);
   button->tooltip("Mark Completed");
   button->image(images_.get(ImageDB::COMPLETE));
   button->align(FL_ALIGN_IMAGE_NEXT_TO_TEXT);
   button->callback(all_cb, (void*)MARK_COMPLETED_CB);
   button->box(FL_NO_BOX);
   button->down_box(FL_UP_BOX);

   // Toolbar / Mark Canceled
   button=new Fl_Button(button->x()+button->w()+UI::default_field_gap, toolbar->y(), UI::default_field_height, UI::default_field_height);
   button->tooltip("Mark Canceled");
   button->image(images_.get(ImageDB::CANCEL));
   button->align(FL_ALIGN_IMAGE_NEXT_TO_TEXT);
   button->callback(all_cb, (void*)MARK_CANCELED_CB);
   button->box(FL_NO_BOX);
   button->down_box(FL_UP_BOX);

   // Toolbar / Separator & Project
   button=new Fl_Button(button->x()+button->w()+UI::default_field_height/3, toolbar->y(), UI::default_field_height, UI::default_field_height);
   button->tooltip("Project");
   button->image(images_.get(ImageDB::PROJECT));
   button->align(FL_ALIGN_IMAGE_NEXT_TO_TEXT);
   button->callback(all_cb, (void*)VIEW_PROJECT_CB);
   button->box(FL_NO_BOX);
   button->down_box(FL_UP_BOX);

   // Toolbar / Person
   button=new Fl_Button(button->x()+button->w()+UI::default_field_gap, toolbar->y(), UI::default_field_height, UI::default_field_height);
   button->tooltip("Person");
   button->image(images_.get(ImageDB::PERSON));
   button->align(FL_ALIGN_IMAGE_NEXT_TO_TEXT);
   button->callback(all_cb, (void*)VIEW_PERSON_CB);
   button->box(FL_NO_BOX);
   button->down_box(FL_UP_BOX);

   // Toolbar / History
   button=new Fl_Button(button->x()+button->w()+UI::default_field_gap, toolbar->y(), UI::default_field_height, UI::default_field_height);
   button->tooltip("History");
   button->image(images_.get(ImageDB::CLOCK));
   button->align(FL_ALIGN_IMAGE_NEXT_TO_TEXT);
   button->callback(all_cb, (void*)VIEW_HISTORY_CB);
   button->box(FL_NO_BOX);
   button->down_box(FL_UP_BOX);

   // Tollbar / Statusbar
   sb_=new Fl_Box(button->x()+button->w()+UI::default_field_height/3, toolbar->y(), UI::default_window_width-(button->x()+button->w()+UI::default_field_height/3), UI::default_field_height);
   sb_->align(FL_ALIGN_RIGHT|FL_ALIGN_INSIDE);

   // Toolbar final setup
   toolbar->resizable(sb_);
   toolbar->end();

   // Content - Task list
   list_=new TaskListView(0, toolbar->y()+toolbar->h()+UI::default_field_gap, UI::default_window_width, DWH-(toolbar->y()+toolbar->h()+UI::default_field_gap));
   list_->callback(all_cb, (void*)LIST_CB);
   list_->show();

   // Window final setup
   ui_.set_icon(window_);
   window_->resizable(list_);
   window_->end();
   window_->show(argc, argv);
   ui_.init();
   window_->callback(all_cb, this);
   window_->size_range(350, 150);

   set_show_all(KeyDAO::get_dao().current(KeyDAO::SHOW_ALL_KEY));
   show_all();
   ui_.set_color_menu(menubar_); // this must be called after ui_.init();
}

void
TaskUI::all_cb(Fl_Widget* widget, void* data)
{
   TaskUI* ui=0;
   int option=(long)data;

   if(!widget->window())
   {
      ui=(TaskUI*)widget->user_data();
      option=QUIT_CB;
   }
   else
   {
      ui=(TaskUI*)widget->window()->user_data();
   }

   switch(option)
   {
      case ABOUT_CB:
         ui->ui_.about();
         break;
      case ADD_CB:
         ui->add();
         break;
      case COLOR_AMBIANCE_CB:
         ui->ui_.set_theme(UI::AMBIANCE_THEME);
         break;
      case COLOR_DARK_CB:
         ui->ui_.set_theme(UI::DARK_THEME);
         break;
      case COLOR_DEFAULT_CB:
         ui->ui_.set_theme(UI::DEFAULT_THEME);
         break;
      case COLOR_NUMIX_CB:
         ui->ui_.set_theme(UI::NUMIX_THEME);
         break;
      case EDIT_CB:
         ui->edit();
         break;
      case LIST_CB:
         if(Fl::event_clicks() || Fl::event_key() == FL_Enter)
         {
            Fl::event_clicks(0);
            ui->edit();
         }
         else
         {
            ui->list_->callback_sort();
         }
         break;
      case MARK_CANCELED_CB:
         ui->mark_canceled();
         break;
      case MARK_COMPLETED_CB:
         ui->mark_completed();
         break;
      case QUIT_CB:
         ui->ui_.quit();
         break;
      case SHOW_ALL_CB:
         ui->show_all();
         break;
      case VIEW_HISTORY_CB:
         ui->ui_.view_occurrence();
         break;
      case VIEW_PERSON_CB:
         ui->ui_.view_person();
         break;
      case VIEW_PROJECT_CB:
         ui->ui_.view_project();
         break;
   }
}

void
TaskUI::show()
{
   list_->take_focus();
   window_->show();
}

void
TaskUI::hide()
{
   window_->hide();
}

void
TaskUI::redraw()
{
   if(window_->shown())
   {
      window_->redraw();
      UI::set_color_menu(menubar_);
   }
}

const Fl_Window*
TaskUI::get_window() const
{
   return window_;
}

void
TaskUI::add()
{
   ui_.edit_task();
}

void
TaskUI::edit()
{
   Task* task=list_->get_selected();
   if(task != 0)
   {
      ui_.edit_task(task);
   }
}

void
TaskUI::mark_canceled()
{
   ui_.mark_task_canceled(list_->get_selected());
}

void
TaskUI::mark_completed()
{
   ui_.mark_task_completed(list_->get_selected());
}

Fl_Menu_Item*
TaskUI::get_show_all_menu_item()
{
   return (Fl_Menu_Item*)menubar_->find_item("&View/Show &All        ");
}

bool
TaskUI::get_show_all()
{
   Fl_Menu_Item* item=get_show_all_menu_item();
   if(item)
   {
      return item->value();
   }
}

void
TaskUI::set_show_all(const bool& value)
{
   Fl_Menu_Item* item=get_show_all_menu_item();
   if(item)
   {
      if(value)
      {
         item->set();
      }
      else
      {
         item->clear();
      }
   }
}

void
TaskUI::show_all()
{
   bool selected=get_show_all();
   KeyDAO::get_dao().set(KeyDAO::SHOW_ALL_KEY, selected);
   list_->show(selected);
   update_statusbar();
}

void
TaskUI::update_statusbar()
{
   bool show_all=get_show_all();
   std::stringstream ss;
   if(show_all)
   {
      ss << dao_.count() << " total";
   }
   else
   {
      ss << dao_.count_opened() << " opened";
   }
   sb_->copy_label(ss.str().c_str());
}

void
TaskUI::notify(const Subject& subject __attribute__((unused)),
      const std::string& message __attribute__((unused)))
{
   if(window_->visible_r())
   {
      update_statusbar();
   }
}
